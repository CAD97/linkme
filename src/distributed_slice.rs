use core::mem;
use core::ops::Deref;
use core::slice;

use crate::private::Slice;

pub struct DistributedSlice<T: ?Sized + Slice> {
    start: StaticPtr<T::Element>,
    stop: StaticPtr<T::Element>,
}

struct StaticPtr<T> {
    ptr: *const T,
}

unsafe impl<T> Send for StaticPtr<T> {}

unsafe impl<T> Sync for StaticPtr<T> {}

impl<T> Copy for StaticPtr<T> {}

impl<T> Clone for StaticPtr<T> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<T> DistributedSlice<[T]> {
    #[doc(hidden)]
    pub const unsafe fn private_new(start: *const T, stop: *const T) -> Self {
        DistributedSlice {
            start: StaticPtr { ptr: start },
            stop: StaticPtr { ptr: stop },
        }
    }

    #[doc(hidden)]
    #[inline]
    pub unsafe fn private_typecheck(self, element: T) {
        mem::forget(element);
    }
}

impl<T> DistributedSlice<[T]> {
    pub fn static_slice(self) -> &'static [T] {
        let stride = mem::size_of::<T>();

        if stride == 0 {
            // We could make this work by storing a 1-byte companion entry in a
            // different link_section and using that count as the len if anyone
            // requires this to work, but for now just:
            return &[];
        }

        let start = self.start.ptr;
        let stop = self.stop.ptr;
        let byte_offset = stop as usize - start as usize;
        let len = byte_offset / stride;
        unsafe { slice::from_raw_parts(start, len) }
    }
}

impl<T> Copy for DistributedSlice<[T]> {}

impl<T> Clone for DistributedSlice<[T]> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<T: 'static> Deref for DistributedSlice<[T]> {
    type Target = [T];
    fn deref(&self) -> &'static Self::Target {
        self.static_slice()
    }
}

impl<T: 'static> IntoIterator for DistributedSlice<[T]> {
    type Item = &'static T;
    type IntoIter = slice::Iter<'static, T>;
    fn into_iter(self) -> Self::IntoIter {
        self.static_slice().iter()
    }
}
