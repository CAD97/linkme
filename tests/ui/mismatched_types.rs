use linkme::distributed_slice;

pub struct Bencher;

#[distributed_slice]
pub static BENCHMARKS: [fn(&mut Bencher)] = [..];

#[distributed_slice(BENCHMARKS)]
static BENCH_WTF: usize = 999;

fn main() {}
