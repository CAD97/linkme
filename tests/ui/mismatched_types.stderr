error[E0308]: mismatched types
  --> $DIR/tests/ui/mismatched_types.rs:12:1
   |
12 | fn wrong_bench_fn<'a>(_: &'a mut ()) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `Bencher`, found `()`
   |
   = note: expected fn pointer `for<'r> fn(&'r mut Bencher)`
              found fn pointer `for<'a> fn(&'a mut ())`

error[E0308]: mismatched types
 --> $DIR/tests/ui/mismatched_types.rs:9:19
  |
9 | static BENCH_WTF: usize = 999;
  |                   ^^^^^ expected fn pointer, found `usize`
  |
  = note: expected fn pointer `for<'r> fn(&'r mut Bencher)`
                   found type `usize`
